aries_cloudagent/connections/models/diddoc_v2/diddoc.py

```diff
@@ -239,11 +268,99 @@ def id(self, value: str) -> None:
 
         self._id = value
 
-    def set(
-        self,
-        item: Union[Service, VerificationMethod],
-        upsert=False,
-        verification_type="publicKey",
+    def add_verification_method(
+            self,
+            type: Union[PublicKeyType, str],
+            controller: Union[str, Sequence],
+            value: str,
+            *,
+            ident: str = None,
+            usage: str = None,
+            authentication: bool = False,
+            verification_type: str = "publicKey",
+            upsert: bool = False
+    ) -> VerificationMethod:
+        """Add a verification method to this document."""
+        if ident:
+            id = "{}#{}".format(self.id, ident)
+            if self._index.get(id) and not upsert:
+                raise ValueError("ID already exists, use arg upsert to update it")
+        else:
+            for index in range(1, 100):
+                id_aux = "{}#keys-{}".format(self.id, index)
+                if not self._index.get(id_aux):
+                    id = id_aux
+                    break
+
+        key = VerificationMethod(id, type, controller, usage, value, authentication)
+        self._set(key, upsert=upsert, verification_type=verification_type)
+        return key
+
+    def add_service(
+            self,
+            type: Union[str, list],
+            endpoint: Union[str, Sequence, dict],
+            ident: str = None,
+            priority: int = 0,
+            upsert: bool = False
+    ) -> Service:
+        """Add service to this document."""
+        if ident:
+            id = "{}#{}".format(self.id, ident)
+            if self._index.get(id) and not upsert:
+                raise ValueError("ID already exists, use arg upsert to update it")
+        else:
+            for index in range(1, 100):
+                id_aux = "{}#service-{}".format(self.id, index)
+                if not self._index.get(id_aux):
+                    id = id_aux
+                    break
+
+        service = Service(id, type, endpoint, priority=priority)
+        self._set(service, upsert=upsert)
+
+    def _add_keys(self, keys):
+        result = []
+        if isinstance(keys, list):
+            for key in keys:
+                result.append(key.id)
+                self._set(key)
+
+        elif isinstance(keys, VerificationMethod):
+            result.append(keys.id)
+            self._set(keys)
+
+        return result
+
+    def add_didcomm_service(
+            self,
+            type: Union[str, list],
```

Comment:
For our purposes, type should be optional, defaulting to `did-communication` when not specified. For backwards compatibility, this method should produce two services, one with `did-communication` as the type and one with `IndyAgent` as the type, and add both to the document. If the caller needs more control over this behavior, they should revert to the previous mechanism of directly constructing a service object and adding it to the document manually or use the generic `add_service` method.